### PR TITLE
sub: add --(sub/osd)-border-style & refactoring option naming to match ASS spec

### LINKED ISSUE #14194: Make sub-back-color follow text precisely
### Expected behavior of the wanted feature

I'd like that MPV's _sub-back-color_ looked like MPC-HC's _opaque box_, where the text of a shorter line is precisely covered by the back color, without the need of the box being a perfect rectangle:
Screenshot (188)

On MPV on the other hand, it currently looks like this:
Bakemonogatari 13 - Tsubasa Cat, Part 3 mkv_snapshot_20 13 921 【MPV】【LINES】 (MTBB (Honorifics))

The _sub-back-color_ feature is important to me because I'm visually impaired and use it to improve subtitles' readability, but the way it looks like on MPV is not the ideal I suppose.
Btw, I don't want the current behavior to be completely withdrawn, just that an option to make this behavior customizable existed (_sub-back-precise=yes_ maybe? I can't think of a better name at the moment).

### Alternative behavior of the wanted feature

_No response_

### Log File

_No response_

### Sample Files

_No response_

### PR DESCRIPTION
**Pull request by ruihe774**
that draw background that tightly wraps each line of text.

Fixes #14194

### CODE REVIEW

**Comment by ruihe774** [comment]
@KonoVitoDa plz test it.

**Comment by KonoVitoDa** [comment]
Works perfectly, thank you very much!

**Comment by hooke007** [comment]
close https://github.com/mpv-player/mpv/issues/10464 instead

**Review by kasper93 — APPROVED** [review]

**Comment by llyyr** [comment]
Can we also switch to tight by default?

**Comment by ruihe774** [comment]
> Can we also switch to tight by default?

What's the point to make a breaking change? IMO this option is only related to personal preference. Users can easily set this option by themselves.

**Comment by KonoVitoDa** [comment]
How do I get the updated version with the new feature implemented?

**Comment by ruihe774** [comment]
> How do I get the updated version with the new feature implemented?

Before this is merged, you can get the build artifacts from https://github.com/mpv-player/mpv/actions/workflows/build.yml. Keep in mind you need to find build artifacts of correct PR.

After this is merged, you can get it from unofficial nightly builds https://mpv.io/installation/

**Comment by llyyr** [comment]
> What's the point to make a breaking change? IMO this option is only related to personal preference. Users can easily set this option by themselves.

Users can change many options by themselves, but there's still value in the default value being something that doesn't need to be changed for most users. I'd argue the current bounding box is not wanted by anyone.

**Comment by ruihe774** [comment]
> > What's the point to make a breaking change? IMO this option is only related to personal preference. Users can easily set this option by themselves.
> 
> Users can change many options by themselves, but there's still value in the default value being something that doesn't need to be changed for most users. I'd argue the current bounding box is not wanted by anyone.

Worth noting that BorderStyle=3 has suboptimal behavior for non-opaque colors. e.g.:
BorderStyle=3

I guess it was the reason mpv chose 4 as the default.

**Comment by KonoVitoDa** [comment]
Will this update also change the way ```--sub-back-color``` is selected? Because on my mpv.config it's set to ```#000000``` (black), but in this specific part it seems to be using the sub's original border color:
Erai-raws  Sousou no Frieren - 01  1080p 81106F0B mkv_snapshot_01 31 425 【MPV】【LINES】 (Portuguese(Brazil))

Erai-raws  Sousou no Frieren - 01  1080p 81106F0B mkv_snapshot_01 31 466 【MPV】【LINES】 (Portuguese(Brazil))

**Comment by ruihe774** [comment]
> Will this update also change the way `--sub-back-color` is selected? Because on my mpv.config it's set to `#000000` (black), but in this specific part it seems to be using the sub's original border color: Erai-raws Sousou no Frieren - 01 1080p 81106F0B mkv_snapshot_01 31 425 【MPV】【LINES】 (Portuguese(Brazil))
> 
> Erai-raws Sousou no Frieren - 01 1080p 81106F0B mkv_snapshot_01 31 466 【MPV】【LINES】 (Portuguese(Brazil))

The behavior did not change. It only applies to SRT subtitles. If you want to override ASS subtitles that select a BackColor itself , you may use `--sub-ass-style-overrides`.

**Comment by KonoVitoDa** [comment]
> ASS subtitles that select a BackColor itself

But this same subtitle is shown with the black back color in the old version, so it seems to be something related to the new feature:
Erai-raws  Sousou no Frieren - 01  1080p 81106F0B mkv_snapshot_01 43 437 【MPV】【LINES】 (Portuguese(Brazil))

**Comment by ruihe774** [comment]
> > ASS subtitles that select a BackColor itself
> 
> But this same subtitle is shown with the black back color in the old version, so it seems to be something related to the new feature: Erai-raws Sousou no Frieren - 01 1080p 81106F0B mkv_snapshot_01 31 508 【MPV】【LINES】 (Portuguese(Brazil))

Please attach your log file and subtitle files (you can extract it using ffmpeg) so I can look into it.

**Comment by ruihe774** [comment]
I cannot reproduce it. With the same commit of mpv and same set of subtitle related options in your `output-old.txt`, the BackColor and BorderStyle are also selected by ASS subtitle file itself.

Screenshot:
Screenshot from 2024-05-22 16-43-43

My log:
[mpv.log](https://github.com/mpv-player/mpv/files/15400671/mpv.log)

My options:
```
[   0.001][v][cplayer] Setting option 'config' = 'no' (flags = 8)
[   0.001][v][cplayer] Setting option 'sub-files-append' = 'Erai-raws.Sousou.no.Frieren.-.01.1080p.81106F0B._Subtitles02.POR/[Erai-raws] Sousou no Frieren - 01 [1080p][81106F0B]_Subtitles02.POR.ass' (flags = 8)
[   0.001][v][cplayer] Setting option 'sub-back-color' = '#000000' (flags = 8)
[   0.001][v][cplayer] Setting option 'sub-border-color' = '#000000' (flags = 8)
[   0.001][v][cplayer] Setting option 'sub-ass-vsfilter-aspect-compat' = 'yes' (flags = 8)
[   0.001][v][cplayer] Setting option 'sub-scale-with-window' = 'no' (flags = 8)
[   0.001][v][cplayer] Setting option 'sub-ass-scale-with-window' = 'yes' (flags = 8)
[   0.001][v][cplayer] Setting option 'sub-font-size' = '53' (flags = 8)
[   0.001][v][cplayer] Setting option 'sub-pos' = '100' (flags = 8)
[   0.001][v][cplayer] Setting option 'sub-margin-x' = '25' (flags = 8)
[   0.001][v][cplayer] Setting option 'sub-margin-y' = '22' (flags = 8)
[   0.001][v][cplayer] Setting option 'sub-use-margins' = '' (flags = 8)
[   0.001][v][cplayer] Setting option 'osd-font-size' = '25' (flags = 8)
[   0.001][v][cplayer] Setting option 'osd-color' = '#FFFFFFFF' (flags = 8)
[   0.001][v][cplayer] Setting option 'osd-back-color' = '#000000' (flags = 8)
[   0.001][v][cplayer] Setting option 'log-file' = '/var/data/mpv.log' (flags = 8)
```

**Comment by KonoVitoDa** [comment]
Really weird. Is not it related to the used FFmpeg version?
Btw, I don't know why the FFmpeg version on my output-old.txt is shown as ```N-115357-g0c1304ae1``` while I have ```ffmpeg version 2024-05-20-git-127ded5078-full_build-www.gyan.dev``` on my PATH.
mpv's updater also says FFmpeg doesn't exist (I downloaded my build from [here](https://sourceforge.net/projects/mpv-player-windows/files/64bit/)).

**Comment by ruihe774** [comment]
> Btw, I don't know why the FFmpeg version on my output-old.txt is shown as N-115357-g0c1304ae1 while I have ffmpeg version 2024-05-20-git-127ded5078-full_build-www.gyan.dev on my PATH.

mpv's CI statically links with ffmpeg. It's not overridable. Many unofficial builds also do that.

> Really weird. Is not it related to the used FFmpeg version?

I don't think it is related to ffmpeg version.

**Comment by KonoVitoDa** [comment]
Oh, I was sleepy and forgot to mention that I use this keybind to switch between sub's original style and my own style:
```d	cycle-values sub-ass-override "force" "no"```

Both of the screenshots I shared before where with `sub-ass-override` set to `force`, but I don't understand why on _mpv-x86_64-20240522-git-0dd6321_ it made the back color black while on the artifact version _[mpv-x86_64-w64-mingw32](https://nightly.link/mpv-player/mpv/actions/artifacts/1521477767.zip)_ it used original sub's border color.

**Comment by ruihe774** [comment]
> Oh, I was sleepy and forgot to mention that I use this keybind to switch between sub's original style and my own style: `d cycle-values sub-ass-override "force" "no"`
> 
> Both of the screenshots I shared before were with `sub-ass-override` set to `force`, but I don't understand why on _mpv-x86_64-20240522-git-0dd6321_ it made the back color black while on the artifact version _[mpv-x86_64-w64-mingw32](https://nightly.link/mpv-player/mpv/actions/artifacts/1521477767.zip)_ it used original sub's border color.

Reproduced with `--sub-ass-override=force`.

Screenshot from 2024-05-23 01-35-51

**Comment by KonoVitoDa** [comment]
Well, if this only occurs with specific subtitle lines (this one is a typesetting), so I think it's not a big problem. Actually it's even better to have a totally grey `sub-back-color` than a black `sub-back-color` with a grey `sub-border` over it.

**Comment by KonoVitoDa** [comment]
> Actually it makes sense to have a totally grey `sub-back-color` than a black `sub-back-color` with a grey `sub-border` over it.

But it would be great if this behavior could also be customized by user.

**Comment by ruihe774** [comment]
It is related to the selection of different colors in ASS. BorderStyle=3 will use OutlineColor and BorderStyle=4 will use BackColor. I'm refactoring related options to make it more consistent.

**Comment by ruihe774** [comment]
> For the sample file, when the text border is drawn, it respects the BorderStyle setting and draws the background with the OutlineColour. Even if the text has a non-zero shadow offset, its color is overwritten by the later drawn border color. AFAIK this is expected ASS behavior, while the nonstandard BorderStyle = 4 disregards that. See https://github.com/libass/libass/pull/105#issuecomment-44789530.

> > BorderStyle=3 will use OutlineColor and BorderStyle=4 will use BackColor.


> This is not true, see above. BorderStyle = 3 uses both colors. It's just the shadow color isn't visible when the shadow offset is zero and thus not drawn.

Yeah. OutlineColor, BackColor, and BorderStyle are orthogonal.

**Comment by ruihe774** [comment]
I've made some breaking changes to option naming to match ASS spec. Some option "dispatching" logic has been removed. IMO now it's more simple and intuitive.

**Comment by ruihe774** [comment]
> There is no benefit of the `border -> outline` renaming change. Just add some clarification in the documentation instead. Please revert.

I do not agree.

- It matches the naming in specs.
- `BorderStyle=4` does not use `--sub-border-color`, which leads to confusion.
- It is `OPT_REPLACED`. Existing usages are not broken.
- `--(sub/osd)-shadow-color` is removed. It is already a breaking change. No meaning to continue using the old names for compatibility reasons.

**Comment by ruihe774** [comment]
> > * It matches the naming in specs.
> 
> `border` is a better name for 99% of the cases.
> 
> > * `BorderStyle=4` does not use `--sub-border-color`, which leads to confusion.
> 
> There is no confusion: it's `libass`-specific and not part of the ASS spec. It's not supported by `xy-VSFilter`.
> 
> > * `--(sub/osd)-shadow-color` is removed. It is already a breaking change. No meaning to continue using the old names for compatibility reasons.
> 
> This breaking is not needed when you can alias it to `--sub-back-color`.

OK. I keep them as aliases.

File: sub/osd.c
Comment (ruihe774) [reply sub/osd.c:49]: > This is really ugly IMO, just to work around a deficiency in the option system

Yeah.

> but not sure if it's worth doing it in this PR.

There are existing usages that depend on the current behavior. Not feasible to modify all of them in this PR.

File: DOCS/man/options.rst
Comment (ruihe774) [reply DOCS/man/options.rst:2882]: I don't think the outline is also drawn as opaque boxes. It is still drawn as outline.

Screenshot

**Comment by KonoVitoDa** [comment]
Hi. Any news?

**Comment by ruihe774** [comment]
> Hi. Any news?

My work is finished. Please be patient. People need time to review and test the PR.

**Comment by guidocella** [comment]
I think the 2 commits can be squashed, and

```lua
local has_shadow = mp.get_property('osd-back-color'):sub(2, 3) == '00'
```

in console.lua should become

```lua
local has_shadow = mp.get_property('osd-back-color'):find('box$') == nil
```

**Comment by ruihe774** [comment]
Rebased. I think it's time to merge it. @na-na-hi @kasper93

File: sub/osd.c
Comment (Dudemanguy) [reply sub/osd.c:49]: What's this for specifically? Just option aliasing? Shouldn't, e.g., `{"shadow-color", OPT_ALIAS("sub-back-color")},` already work or am I missing something here?

Comment (ruihe774) [reply sub/osd.c:49]: It can be refreshed later. IMO this should not be a block on this PR.

**Comment by kasper93** [comment]
> Rebased. I think it's time to merge it. @na-na-hi @kasper93

I don't follow all the sub/osd styles, ask someone else to review those.

File: sub/osd.c
Comment (Dudemanguy) [inline sub/osd.c:77]: Could you squash this change in the first commit?

**Comment by Dudemanguy** [comment]
>These new option names stick to whatever names the ASS format uses, but as pointed out, the meanings of these names can be completely different depending on the border style.

If what happens changes completely depending on the border style, I think we're just doomed, no? I don't have any good name ideas anyway.

File: DOCS/man/options.rst
Comment (ruihe774) [reply DOCS/man/options.rst:2882]: > I'm not sure if we should hide this peculiarity behind a dedicated margin option or document it as is.

IMO we should be close to the upstream. Hiding something or making an extra layer of abstraction will make things more complicated.

File: sub/ass_mp.c
Comment (ruihe774) [reply sub/ass_mp.c:72]: IMO it's too smart. I consider these options orthogonal.

**Comment by ruihe774** [comment]
IMO 1:1 mappings from CLI options to upstream's options are intuitive. Despite historical reasons, ASS is designed to be in that shape, and users can consult its spec by themselves. Using a complex option handling and mapping logic will make things complicated and hard to understand, unless we thoroughly describe our logic in the documentation, which is a maintain burden and a reinvention of existing spec.

**Comment by ruihe774** [comment]
> What should be done with `--(sub/osd)-border-style=none`? It still doesn't work as documented. This needs to be fixed or dropped as a choice.

Sorry, I forgot it.

It turns out `0` is not a valid value for `BorderStyle` in ASS. Now I just remove it.

File: DOCS/man/options.rst
Comment (Dudemanguy) [inline DOCS/man/options.rst:2863]: Should remove `none` from here.

File: DOCS/man/options.rst
Comment (Dudemanguy) [inline DOCS/man/options.rst:2864]: Probably worth specifying the default value.

File: sub/osd.c
Comment (Dudemanguy) [inline sub/osd.c:47]: No need for a newline here.

File: sub/osd.c
Comment (Dudemanguy) [inline sub/osd.c:88]: Since `osd-back-color` and `osd-shadow-color` are the same now, this causes the back color of the osd bar to not be fully transparent anymore.

**Review by Dudemanguy — CHANGES_REQUESTED** [review]
Sorry for the delay. Some minor things.

File: DOCS/man/options.rst
Comment (Dudemanguy) [inline DOCS/man/options.rst:4470]: Should remove `none` from here too.

File: sub/osd_libass.c
Comment (Dudemanguy) [inline sub/osd_libass.c:428]: This makes the back color not work with `BorderStyle=1` at all. Is that the intention? I just assumed the default for `osd-back-color` would be changed to be fully transparent so it could otherwise be configurable.

Comment (ruihe774) [reply sub/osd_libass.c:428]: With `BorderStyle=1`, `--osd-back-color` means the shadow color. If users want a background color, they need to set `--osd-border-style=background-box` as well.

Comment (Dudemanguy) [reply sub/osd_libass.c:428]: Sure. Just wanted to make sure that made sense.

**Review by Dudemanguy — APPROVED** [review]
Seems to be ready to go now.

**Comment by KonoVitoDa** [comment]
Hi. Is it implemented already? I couldn't find the new options in the manual.

**Comment by Dudemanguy** [comment]
No it hasn't been merged yet.

**Pull request merged**
